xGov 178: SIWA - Sign In With Algorand

headline-design opened the pull request.

nonfungibleab commented on Proposals/xgov-178.md:
```diff
@@ -0,0 +1,87 @@
+- Full documentation site
+- NPM package publication
+- Production integrations
+
```
Please provide public URL of where/how the community can check proposal deliverables.

headline-design replied:
Absolute. Are you sating to add the url on line 75? or where exactly? I don't want to put it in the wrong place.

headline-design replied:
The work on SIWA is currently being done here - https://github.com/headline-design/xballot/tree/main/packages/siwa. As soon as I have a working prototype I'll create a seperate repository for SIWA.

headline-design replied:
Thrilled to announce that we now have a working prototype. Will share the Github repos, etc. in the coming days.

SudoWeezy commented:
If your proposal is ready, please update the status to `Final` before the 6th of May 2024.

emg110 commented:
IMHO, sign in with Algorand (SIWA) or any other attempt to replicate ERC-4361 for Algorand should not be attempted as implementation or tooling libraries but instead should be an ARC contribution. Also it is notable that ARCs and ERCs are not paid subjects. 
On the other hand we already have ARC31 for such auth use-cases and  ARC60 for arbitrary data/message signing, as @ehanoc mentioned. 
What is the real use-case or added value of this limited (to react/nextjs only) and very opinionated implementation?

headline-design commented:
SIWA Deliverables (6/18/2024): 

SIWA-NPM: https://www.npmjs.com/package/@avmkit/siwa
SIWA-Github: https://github.com/headline-design/avmkit/tree/main/packages/siwa

CSR-Template-deployed: https://algostack.vercel.app
CSR-Template-Github: https://github.com/headline-design/algostack

SSR-Template-deployed: https://algostack-ssr.vercel.app
SSR-Template-Github: https://github.com/headline-design/algostack-ssr

SIWA-website-https://siwa.org
SIWA-docs: https://siwa.org/help

SIWA-template-marketplace: https://siwa.org/templates

ehanoc commented:
@headline-design I have concerns regarding your delivery. 

- You mention on your proposal "Enterprise Grade" although: 
   - You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing. 
  - No type of testing all (unit, integration, e2e) 

- "EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."

  - The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. **This is not demonstrated**

- "Production integrations":
  - Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this. 

Looking forward for your reply. Thanks!

PhearZero commented:
@headline-design I'm a bit confused, I agree with @emg110 and @ehanoc on this one

I look forward to understanding this better, is there a thread that has more context about this proposal? The original proposal references [docs.login.xyz](https://docs.login.xyz) which already outlined the Next.js integration path. I can't seem to find anywhere in the proposal that mentions the dependency on a dedicated service. I also don't see any guides on how to use the main deliverable, the `@headline-design/siwa` (now `@avmkit`) package nor any of it's [accompanying packages](https://forum.algorand.org/t/xgov-178-siwa-sign-in-with-algorand/11705/3) in a Next.js/React context:

>You can track my progress here - [siwa](https://github.com/headline-design/xballot/tree/main/packages/siwa) and here - [use-siwa](https://github.com/headline-design/xballot/tree/main/apps/atlas/use-siwa). I’ll be moving the code to a SIWA-specific repo once I have a working prototype.

Few questions:

What was the decision process around creating a bespoke template and hosting platform? The template and service includes other opinionated decisions which don't seem directly related to SIWA. What was the decision behind not integrating with [NextAuth](https://docs.login.xyz/integrations/nextauth.js)(as recommended by SIWE) and not providing the packages as the deliverables (similar to WAGMI and what was relayed in the forum)? 

Another question is why does the integration path require registering with a centralized application? SIWA Dashboard does not seem functional, currently throws 500 when using Pera on Step 1 using the guide [illustrated here](https://siwa.org/help/article/how-to-use-siwa-nextjs#how-to-create-a-project). I was able to bypass the error to explorer the product but the connect experience seems broken currently. 


Related: ⚠️ The github integration to SIWA Dashboard is asking for some alarming permissions

![Screenshot from 2024-08-28 14-37-42](https://github.com/user-attachments/assets/ead56d88-4d29-4e58-b0b1-90aade1066db) 

With all due respect, it seems like the SIWA scope has creeped to be an alternative to `algokit` (and possibly Vercel?) with the recent refactor to `avmkit` and the dependencies on the SIWA app/algostack templates. There has clearly been a lot of effort put into this delivery, just wish the incentives aligned with the broader ecosystem. Appreciate the design work with radix, clearly a talented team!

SilentRhetoric commented:
Hi @headline-design, the Foundation is working on reviewing the technical deliverables for this xGov grant. In order for this review to proceed, please provide the following:

1. Clicking on the deliverables link in the original proposal [https://github.com/headline-design/siwa](https://github.com/headline-design/siwa) results in a 404 error. Can you please remediate this in one of two ways? A) Retain the original link in the proposal and ensure that it works, or B) submit a PR to amend the link in the proposal so that it correctly links to the URL of the deliverables.

2. The SIWA-NPM package at [https://www.npmjs.com/package/@avmkit/siwa](https://www.npmjs.com/package/@avmkit/siwa) / the underlying repository [https://github.com/headline-design/avmkit/tree/main/packages/siwa](https://github.com/headline-design/avmkit/tree/main/packages/siwa) should have a proper README file that explains the package to visitors to the NPM package and the repository itself. I would expect an open-source library submitted under the Tools category of xGov to contain instructions for how it can be used by developers.

3. Please update the deliverables per instructions in the xGov legal contract section 5.5(g):
> ...you acknowledge and credit us by identifying the phrase “This work has been performed with support from the Algorand Foundation xGov Grants Program” in a prominent area in all forms of outputs in relation to the Deliverables and all other activities sponsored under these Terms

4. Please provide answers to the three questions articulated by @ehanoc in this comment [https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2315795843](https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2315795843)

5. Please provide answers to the four questions articulated by @PhearZero in this comment [https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2316137518](https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2316137518)

We look forward to your response!

headline-design commented:
@ehanoc 

You mention on your proposal "Enterprise Grade" although:

You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing.
No type of testing all (unit, integration, e2e)

Answer. "Enterprise Grade" for the purpose of this proposal is a reference to traffic throughput and Enterprise product compatability. Enterprise meaning authentication that functions in parallel to standard Enterprise authentication providers such as Google oAuth. Enterprise for the purpose of this proposal is not referencing an opinionated security standard. SIWA is enterprise-grade in the same way that SIWE (Sign-In-With-Ethereum) is enterprise-grade or SIWS (Sign-In-With-Substrate) is enterprise-grade.

"EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."

The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. This is not demonstrated

Answer. The proposal does not list in deliverables an example or examples of evm-compatability. The focus of this proposal is on providing the building blocks for this compatability. However, if you look at the template examples and compare it to our related Dotstack template for Polkadot with SIWS - Sign In With Substrate, you will start to see the connective tissue for our goal of creating a true multi-vm auth standard with SIWE, SIWA, SIWS, and SIWS - Sign In With Solana. For example one of the starter templates could be modified to support SIWE in maybe 1-2 hours and I'd be happy to make that template as part of the next xGov session. Basically, we have already built out the supporting hooks libraries for SIWE, and SIWS to support Algorand sign-in compatibility with Substrate and Ethereum. Our hooks library work with Substrate for example won the Consensus Hackathon Polkadot track and Talisman Wallet side track in June.

"Production integrations":

Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this.
Looking forward for your reply. Thanks! 

Answer: SIWA - Sign In With Algorand is live in production at https://siwa.org, https://hdl.world, https://voiager.org, and our starter templates at https://algostack-ssr.vercel.app, and https://algostack.vercel.app. Total users for the Atlas framework which powers hdl.world and voiager.org just hit 2600 users. You can check out all of our Atlas stats here - https://hdl.world/atlas/analytics. The majority of users signed up with SIWA authentication and of those over 1500 users created new wallets through the xWallet and SIWA flow.

headline-design commented:
@PhearZero 

Few questions:

What was the decision process around creating a bespoke template and hosting platform for SIWA? The template and service includes other opinionated decisions which don't seem directly related to SIWA. What was the decision behind not publishing the required [Auth.js](https://docs.login.xyz/integrations/nextauth.js) provider package as a part of the deliverables (similar to WAGMI+siwe and what appeared to be relayed in the forum)?

Answer 1. It's important to seperate the SIWA auth package from semi-related elements. The SIWA auth package is framework-agnostic and will work with React, Next, Solid, Svelte, etc. The idea behind the SIWA website dev platform really derived from the idea that there should be a proper explorer for SIWA templates. It's important to note that the SIWA website is not a hosting platform, it's a doc site and template explorer. The user login part of the SIWA website allows users to create and publish their own SIWA templates that can hopefully evolve into a robust template explorer for Algorand. It's also important to note that the user template publisher is not part of the proposal, has not been thoroughly tested, and is meant to be more of a demo for what we would like SIWA to become as a hub for Algorand-based templates. 

Another question is why does the integration path require registering with a centralized application? SIWA Dashboard does not seem functional, currently throws 500 when using Pera on Step 1 using the guide [illustrated here](https://siwa.org/help/article/how-to-use-siwa-nextjs#how-to-create-a-project). I was able to bypass the error to explore the product but the connect experience seems broken currently.


Answer 2. Right this goes back to what I said previously, the SIWA dashboard/vs the docs part of the site is more of a prototype - not part of SIWA deliverables. The main motiviation behind building it out like that is that the Algorand community will see that there is an exciting roadmap for SIWA beyond just a one-off proposal with no future development. If it would be prefereble for you, we can disable user login on the siwa.org site for now, as I mentioned its more of a prototype for now. 


2.1 One idea I had was using part of the proposal funds to create a bunch of Hackathon bounties for publishing templates with SIWA. Like you mentioned starting on a use-wallet SIWA template. that would be a perfect example of a bounty template that could be published on the SIWA template explorer. 

Related: ⚠️ The github integration to SIWA Dashboard is asking for some alarming permissions

Screenshot from 2024-08-28 14-37-42

With all due respect, it seems like the SIWA scope has creeped to be an alternative to algokit (and possibly Vercel?) with the recent refactor to avmkit and the dependencies on the SIWA app/algostack templates. There has clearly been a lot of effort put into this delivery, just wish the incentives aligned with the broader ecosystem. Appreciate the design work with radix, clearly a talented team!

Answer 3. Its important to note that SIWA is not for authoring smart contracts, it's for  connecting to Dapps and creating UI/UX for Dapps, so no real overlap with algokit. It's also not for hosting anything, the SIWA dashboard is for publishing and exploring SIWA templates while the main site is for documentation. On the macro scale, I believe this is what makes it align really nicely with the broader ecosystem. The SIWA package is an agnostic auth standard. The future SIWA dashboard will allow users to explore and publish (and hopefully be rewarded) for creating awesome templates for the Algorand ecosystem.

headline-design commented:
@PhearZero just to bump this up. The SIWA dashboard is a prototype and hasn't been thoroughly tested and most definetly has bugs, etc. For that reason do you recomend we disable user sign in for now? Since it's still technically a demo of our future plans and not technically part of the current proposal?

ehanoc commented:
> @ehanoc
> 
> You mention on your proposal "Enterprise Grade" although:
> 
> You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing. No type of testing all (unit, integration, e2e)
> 
> Answer. "Enterprise Grade" for the purpose of this proposal is a reference to traffic throughput and Enterprise product compatability. Enterprise meaning authentication that functions in parallel to standard Enterprise authentication providers such as Google oAuth. Enterprise for the purpose of this proposal is not referencing an opinionated security standard. SIWA is enterprise-grade in the same way that SIWE (Sign-In-With-Ethereum) is enterprise-grade or SIWS (Sign-In-With-Substrate) is enterprise-grade.
> 
> "EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."
> 
> The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. This is not demonstrated
> 
> Answer. The proposal does not list in deliverables an example or examples of evm-compatability. The focus of this proposal is on providing the building blocks for this compatability. However, if you look at the template examples and compare it to our related Dotstack template for Polkadot with SIWS - Sign In With Substrate, you will start to see the connective tissue for our goal of creating a true multi-vm auth standard with SIWE, SIWA, SIWS, and SIWS - Sign In With Solana. For example one of the starter templates could be modified to support SIWE in maybe 1-2 hours and I'd be happy to make that template as part of the next xGov session. Basically, we have already built out the supporting hooks libraries for SIWE, and SIWS to support Algorand sign-in compatibility with Substrate and Ethereum. Our hooks library work with Substrate for example won the Consensus Hackathon Polkadot track and Talisman Wallet side track in June.
> 
> "Production integrations":
> 
> Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this. Looking forward for your reply. Thanks!
> 
> Answer: SIWA - Sign In With Algorand is live in production at https://siwa.org, https://hdl.world, https://voiager.org, and our starter templates at https://algostack-ssr.vercel.app, and https://algostack.vercel.app. Total users for the Atlas framework which powers hdl.world and voiager.org just hit 2600 users. You can check out all of our Atlas stats here - https://hdl.world/atlas/analytics. The majority of users signed up with SIWA authentication and of those over 1500 users created new wallets through the xWallet and SIWA flow.

Hey @headline-design ,

I feel that those answers aren't satisfactory. You are claiming after the fact that "we didn't really mean Enterprise Grade", "We didn't really mention EVM compatibility" (which you did) and your production integrations aren't any app or wallet that is used in the ecosystem. 

For this deliverable to be considered successfully deliverable in accordance to your own proposal and value proposition that people voted on, it needs to:

- Refactored to met common Enterprise grade level standards for implementation
- Proven & live examples of EVM compatibility
- Live production examples in known app / wallets of the ecosystem 

Best

johnalanwoods commented:
Also is there a spec on the applied crypto design?

headline-design commented:
> Also is there a spec on the applied crypto design?
> 
> Security is critical on this, needs community review.

Hi John,
Can you ellaborate on the design doc you are looking for. Or can you point me to an example applied crypto design?

headline-design commented:
> > @ehanoc
> > You mention on your proposal "Enterprise Grade" although:
> > You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing. No type of testing all (unit, integration, e2e)
> > Answer. "Enterprise Grade" for the purpose of this proposal is a reference to traffic throughput and Enterprise product compatability. Enterprise meaning authentication that functions in parallel to standard Enterprise authentication providers such as Google oAuth. Enterprise for the purpose of this proposal is not referencing an opinionated security standard. SIWA is enterprise-grade in the same way that SIWE (Sign-In-With-Ethereum) is enterprise-grade or SIWS (Sign-In-With-Substrate) is enterprise-grade.
> > "EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."
> > The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. This is not demonstrated
> > Answer. The proposal does not list in deliverables an example or examples of evm-compatability. The focus of this proposal is on providing the building blocks for this compatability. However, if you look at the template examples and compare it to our related Dotstack template for Polkadot with SIWS - Sign In With Substrate, you will start to see the connective tissue for our goal of creating a true multi-vm auth standard with SIWE, SIWA, SIWS, and SIWS - Sign In With Solana. For example one of the starter templates could be modified to support SIWE in maybe 1-2 hours and I'd be happy to make that template as part of the next xGov session. Basically, we have already built out the supporting hooks libraries for SIWE, and SIWS to support Algorand sign-in compatibility with Substrate and Ethereum. Our hooks library work with Substrate for example won the Consensus Hackathon Polkadot track and Talisman Wallet side track in June.
> > "Production integrations":
> > Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this. Looking forward for your reply. Thanks!
> > Answer: SIWA - Sign In With Algorand is live in production at https://siwa.org, https://hdl.world, https://voiager.org, and our starter templates at https://algostack-ssr.vercel.app, and https://algostack.vercel.app. Total users for the Atlas framework which powers hdl.world and voiager.org just hit 2600 users. You can check out all of our Atlas stats here - https://hdl.world/atlas/analytics. The majority of users signed up with SIWA authentication and of those over 1500 users created new wallets through the xWallet and SIWA flow.
> 
> Hey @headline-design ,
> 
> I feel that those answers aren't satisfactory. You are claiming after the fact that "we didn't really mean Enterprise Grade", "We didn't really mention EVM compatibility" (which you did) and your production integrations aren't any app or wallet that is used in the ecosystem.
> 
> For this deliverable to be considered successfully deliverable in accordance to your own proposal and value proposition that people voted on, it needs to:
> 
> * Refactored to met common Enterprise grade level standards for implementation
> * Proven & live examples of EVM compatibility
> * Live production examples in known app / wallets of the ecosystem
> 
> Best

It's clear you are more interested in arguing semantics than having a good faith discussion about the proposal. To directly reference the proposal deliverables - 

"Our grant application targets one critical milestones:

SIWA Protocol publication: Protocol publication will include full documentation site, NPM package publication, and all relevent integrations for use in production.
Cost Breakdown

Protocol publication - total: (150,000A)

Full documentation site
NPM package publication
Production integrations ." These are explicitly the deliverables for xGov 178. All of which are satisfied. The proposal makes very clear what the scope of the deliverables are, and the in-production examples + templates I have provided actually exceed the deliverables in many key ways. I'd encourage you to refer back to the deliverables sections of proposals before you engage in hostile, bad-faith commentary.

ehanoc commented:
> > > @ehanoc
> > > You mention on your proposal "Enterprise Grade" although:
> > > You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing. No type of testing all (unit, integration, e2e)
> > > Answer. "Enterprise Grade" for the purpose of this proposal is a reference to traffic throughput and Enterprise product compatability. Enterprise meaning authentication that functions in parallel to standard Enterprise authentication providers such as Google oAuth. Enterprise for the purpose of this proposal is not referencing an opinionated security standard. SIWA is enterprise-grade in the same way that SIWE (Sign-In-With-Ethereum) is enterprise-grade or SIWS (Sign-In-With-Substrate) is enterprise-grade.
> > > "EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."
> > > The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. This is not demonstrated
> > > Answer. The proposal does not list in deliverables an example or examples of evm-compatability. The focus of this proposal is on providing the building blocks for this compatability. However, if you look at the template examples and compare it to our related Dotstack template for Polkadot with SIWS - Sign In With Substrate, you will start to see the connective tissue for our goal of creating a true multi-vm auth standard with SIWE, SIWA, SIWS, and SIWS - Sign In With Solana. For example one of the starter templates could be modified to support SIWE in maybe 1-2 hours and I'd be happy to make that template as part of the next xGov session. Basically, we have already built out the supporting hooks libraries for SIWE, and SIWS to support Algorand sign-in compatibility with Substrate and Ethereum. Our hooks library work with Substrate for example won the Consensus Hackathon Polkadot track and Talisman Wallet side track in June.
> > > "Production integrations":
> > > Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this. Looking forward for your reply. Thanks!
> > > Answer: SIWA - Sign In With Algorand is live in production at https://siwa.org, https://hdl.world, https://voiager.org, and our starter templates at https://algostack-ssr.vercel.app, and https://algostack.vercel.app. Total users for the Atlas framework which powers hdl.world and voiager.org just hit 2600 users. You can check out all of our Atlas stats here - https://hdl.world/atlas/analytics. The majority of users signed up with SIWA authentication and of those over 1500 users created new wallets through the xWallet and SIWA flow.
> > 
> > 
> > Hey @headline-design ,
> > I feel that those answers aren't satisfactory. You are claiming after the fact that "we didn't really mean Enterprise Grade", "We didn't really mention EVM compatibility" (which you did) and your production integrations aren't any app or wallet that is used in the ecosystem.
> > For this deliverable to be considered successfully deliverable in accordance to your own proposal and value proposition that people voted on, it needs to:
> > 
> > * Refactored to met common Enterprise grade level standards for implementation
> > * Proven & live examples of EVM compatibility
> > * Live production examples in known app / wallets of the ecosystem
> > 
> > Best
> 
> It's clear you are more interested in arguing semantics than having a good faith discussion about the proposal. To directly reference the proposal deliverables -
> 
> "Our grant application targets one critical milestones:
> 
> SIWA Protocol publication: Protocol publication will include full documentation site, NPM package publication, and all relevent integrations for use in production. Cost Breakdown
> 
> Protocol publication - total: (150,000A)
> 
> Full documentation site NPM package publication Production integrations ." These are explicitly the deliverables for xGov 178. All of which are satisfied. The proposal makes very clear what the scope of the deliverables are, and the in-production examples + templates I have provided actually exceed the deliverables in many key ways. I'd encourage you to refer back to the deliverables sections of proposals before you engage in hostile, bad-faith commentary.

Enterprise Grade, EVM integration and production integrations they aren't semantics, they are or they are not. Currently they are there, or proven. 

Also, from your own proposal :

```
Feature 1: EVM Compatibility
SIWA will be modeled from the ground up to be compatible with SIWE, allowing developers to easily integrate SIWA into their existing Ethereum-based applications. This is a critical feature for developers who build Algorand applications with NextJS. This is a critical feature for developers who build EVM-compatible applications and want to leverage the Algorand blockchain for authentication.

Feature 2: EIP-4361 Equivalent
SIWA will implement an EIP-4361 equivalent on the Algorand blockchain. EI-4361 is an Ethereum Improvement Proposal that defines a standard for decentralized authentication on the Ethereum blockchain. SIWA will provide a similar standard for Algorand, allowing developers to easily integrate decentralized authentication into their applications.

Feature 3: NextJS Support
SIWA will support NextJS, a popular React framework, making it easy for developers to build fast and scalable web applications. NextJS is a powerful tool for building web applications, and SIWA will provide seamless integration with NextJS, allowing developers to build secure and scalable applications with ease.

Feature 4: Wallet/NFD Integration
SIWA will integrate with popular Algorand wallets, popular Algorand libraries, and NFD domains. This integration will allow users to easily sign in to websites and applications using their Algorand wallet, providing a seamless and secure authentication experience.
```

"popular Algorand wallets" : Have you reached out to Pera or Defly to flag that they are willing to adopt this?

johnalanwoods commented:
@ehanoc can you share with @headline-design the paper we wrote on BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace, to give them an example of what I expect.

ehanoc commented:
> @ehanoc can you share with @headline-design the paper we wrote on BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace, to give them an example of what I expect.
> 
> A security description of the protocol is needed.

It's here: https://github.com/algorandfoundation/bip32-ed25519-addendum

headline-design commented:
> > > > @ehanoc
> > > > You mention on your proposal "Enterprise Grade" although:
> > > > You rely on in-memory signing (delegated to the SDK), therefore the secret sees run-time space. An "enterprise grade" solution would isolate key management and signing. No type of testing all (unit, integration, e2e)
> > > > Answer. "Enterprise Grade" for the purpose of this proposal is a reference to traffic throughput and Enterprise product compatability. Enterprise meaning authentication that functions in parallel to standard Enterprise authentication providers such as Google oAuth. Enterprise for the purpose of this proposal is not referencing an opinionated security standard. SIWA is enterprise-grade in the same way that SIWE (Sign-In-With-Ethereum) is enterprise-grade or SIWS (Sign-In-With-Substrate) is enterprise-grade.
> > > > "EVM Compatibility: SIWA will have tight compatibility with SIWE (Sign In With Ethereum), allowing developers to easily integrate SIWA into their existing Ethereum-based applications."
> > > > The SIWA package signs with EdDSA (using algosdk); which is a Ed25519 signature scheme; this would not be compatible with any EVM applications without any form of account abstraction by deploy an EVM contract to deal with the different signature schemes. This is not demonstrated
> > > > Answer. The proposal does not list in deliverables an example or examples of evm-compatability. The focus of this proposal is on providing the building blocks for this compatability. However, if you look at the template examples and compare it to our related Dotstack template for Polkadot with SIWS - Sign In With Substrate, you will start to see the connective tissue for our goal of creating a true multi-vm auth standard with SIWE, SIWA, SIWS, and SIWS - Sign In With Solana. For example one of the starter templates could be modified to support SIWE in maybe 1-2 hours and I'd be happy to make that template as part of the next xGov session. Basically, we have already built out the supporting hooks libraries for SIWE, and SIWS to support Algorand sign-in compatibility with Substrate and Ethereum. Our hooks library work with Substrate for example won the Consensus Hackathon Polkadot track and Talisman Wallet side track in June.
> > > > "Production integrations":
> > > > Which ones? As far as i know, none of the main wallets / apps in production have adopted or integrate this. Looking forward for your reply. Thanks!
> > > > Answer: SIWA - Sign In With Algorand is live in production at https://siwa.org, https://hdl.world, https://voiager.org, and our starter templates at https://algostack-ssr.vercel.app, and https://algostack.vercel.app. Total users for the Atlas framework which powers hdl.world and voiager.org just hit 2600 users. You can check out all of our Atlas stats here - https://hdl.world/atlas/analytics. The majority of users signed up with SIWA authentication and of those over 1500 users created new wallets through the xWallet and SIWA flow.
> > > 
> > > 
> > > Hey @headline-design ,
> > > I feel that those answers aren't satisfactory. You are claiming after the fact that "we didn't really mean Enterprise Grade", "We didn't really mention EVM compatibility" (which you did) and your production integrations aren't any app or wallet that is used in the ecosystem.
> > > For this deliverable to be considered successfully deliverable in accordance to your own proposal and value proposition that people voted on, it needs to:
> > > 
> > > * Refactored to met common Enterprise grade level standards for implementation
> > > * Proven & live examples of EVM compatibility
> > > * Live production examples in known app / wallets of the ecosystem
> > > 
> > > Best
> > 
> > 
> > It's clear you are more interested in arguing semantics than having a good faith discussion about the proposal. To directly reference the proposal deliverables -
> > "Our grant application targets one critical milestones:
> > SIWA Protocol publication: Protocol publication will include full documentation site, NPM package publication, and all relevent integrations for use in production. Cost Breakdown
> > Protocol publication - total: (150,000A)
> > Full documentation site NPM package publication Production integrations ." These are explicitly the deliverables for xGov 178. All of which are satisfied. The proposal makes very clear what the scope of the deliverables are, and the in-production examples + templates I have provided actually exceed the deliverables in many key ways. I'd encourage you to refer back to the deliverables sections of proposals before you engage in hostile, bad-faith commentary.
> 
> Enterprise Grade, EVM integration and production integrations they aren't semantics, they are or they are not. Currently they are there, or proven.
> 
> Also, from your own proposal :
> 
> ```
> Feature 1: EVM Compatibility
> SIWA will be modeled from the ground up to be compatible with SIWE, allowing developers to easily integrate SIWA into their existing Ethereum-based applications. This is a critical feature for developers who build Algorand applications with NextJS. This is a critical feature for developers who build EVM-compatible applications and want to leverage the Algorand blockchain for authentication.
> 
> Feature 2: EIP-4361 Equivalent
> SIWA will implement an EIP-4361 equivalent on the Algorand blockchain. EI-4361 is an Ethereum Improvement Proposal that defines a standard for decentralized authentication on the Ethereum blockchain. SIWA will provide a similar standard for Algorand, allowing developers to easily integrate decentralized authentication into their applications.
> 
> Feature 3: NextJS Support
> SIWA will support NextJS, a popular React framework, making it easy for developers to build fast and scalable web applications. NextJS is a powerful tool for building web applications, and SIWA will provide seamless integration with NextJS, allowing developers to build secure and scalable applications with ease.
> 
> Feature 4: Wallet/NFD Integration
> SIWA will integrate with popular Algorand wallets, popular Algorand libraries, and NFD domains. This integration will allow users to easily sign in to websites and applications using their Algorand wallet, providing a seamless and secure authentication experience.
> ```
> 
> "popular Algorand wallets" : Have you reached out to Pera or Defly to flag that they are willing to adopt this?

You did not support this proposal from the very beginning and I now believe you are maliciously trying to damage it. If you continue in this hostile direction I will look into reporting you to GitHub, the Foundation, etc. I believe what you are doing now is close to online harassment. I don't know who you are or what you have against me but this behavior is completely in appropriate for GitHub. I will no longer be replying to your comments and if you continue atting me in messages I will have to report them. Going forward, if you have actual genuine questions have someone else submit them for you, because I believe you are trying to maliciously damage a proposal you never supported.

headline-design commented:
> @ehanoc can you share with @headline-design the paper we wrote on BIP32-Ed25519 Hierarchical Deterministic Keys over a Non-linear Keyspace, to give them an example of what I expect.
> 
> A security description of the protocol is needed.

Which document in that repo are you referring too?

johnalanwoods commented:
Hey folks, this one https://github.com/algorandfoundation/bip32-ed25519-addendum/blob/main/paper.tex

drichar commented:
@headline-design I took a look at the [source code of SIWA](https://github.com/headline-design/avmkit/blob/main/packages/siwa/src/client.ts) and it contains JSDoc comments that seem odd, referring to the Ethers library and EIPs, e.g.,
```
Creates a parsed Sign-In with Algorand Message (EIP-4361) object
```
```
@param provider Ethers provider to be used for EIP-1271 validation
```

A GitHub global search of one of these comments yields [the original SIWE library](https://github.com/spruceid/siwe/tree/main/packages/siwe/lib) along with several forks, including your https://github.com/headline-design/avmkit repository's SIWA package: https://github.com/search?q=%22This+function+can+be+used+to+retrieve+an+EIP-4361+formated+message%22&type=code

Comparing the two libraries, it is clear that SIWA is ostensibly a fork of SIWE. But there is no mention of SIWE anywhere in the source code. The client.ts file, for example, is virtually identical: https://www.diffchecker.com/iGSuqsk2/

The SIWE library has both Apache 2.0 and MIT licenses, which both require clear attribution in forked versions. The lack of attribution could be interpreted as a violation of the dual license. Even if it isn't, there are obvious ethical considerations since you are seeking a grant for what is largely someone else's work.

SilentRhetoric commented:
As an update, the following matters still need to be addressed by the headline-design team for the technical deliverables review of this grant to advance:

1. A security description of the protocol is needed along the lines of https://github.com/algorandfoundation/bip32-ed25519-addendum/blob/main/paper.tex. _(See https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2381452379.)_
2. Demonstrate EVM compatibility with examples. _(See https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2363148879.)_
3. Demonstrate integration with popular Algorand wallets such a Pera & Defly. _(See https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2363148879.)_
4. Addition of a README to the published NPM package. _(See https://github.com/algorandfoundation/xGov/pull/178#issuecomment-2356720068.)_

headline-design commented:
Just a quick update. We are making a new standalone documentation site that should much better address outstanding questions. Will also get those other notes completed soon. Thanks for your patience!

headline-design commented:
@SilentRhetoric Updated Readme: https://www.npmjs.com/package/@avmkit/siwa

headline-design commented:
The SIWA website will be down while we are setting up the new docs site.

headline-design commented:
@SilentRhetoric Here is the link to the security document. This is my first time working with Latex so I'm happy to update as needed. https://github.com/headline-design/avmkit/blob/main/packages/siwa/security.tex

headline-design commented:
Added an attribution section to the Readme as well. In addition to SIWE, SIWA is also based on the excelent work done by Talisman on Sign In With Substrate and Phantom with Sign In With Solana. All three projects are referenced in the Readme now - https://github.com/headline-design/avmkit/tree/main/packages/siwa

headline-design commented:
@SilentRhetoric Here is the AVM + EVM Template. I used the SSR AlgoStack template so we are calling it Algostack EVM. The main template repo is in avmkit/templates, but I also created a standalone repo for it for forking and it is deployed on Vercel as well. This integrates all the Algostack Algorand stack with Wagmi, Viem, Connectkit for full EVM integration. The prisma database accepts Algorand or EVM wallets as compatible user wallets. Here are the links: Deployed - https://algostack-evm.vercel.app. Standalone repo for forking - https://github.com/headline-design/algostack-evm. Templates repo - https://github.com/headline-design/avmkit/tree/main/templates/algostack-evm

ehanoc commented:
> @SilentRhetoric Here is the link to the security document. This is my first time working with Latex so I'm happy to update as needed. https://github.com/headline-design/avmkit/blob/main/packages/siwa/security.tex

Thanks for the document, but i think this would need to be a technical document. It reads as an article.

Claiming interoperability on a SIWA scheme with EVM / SIWE is a bold endeavour and personally I don't think you demonstrated that. We should be reading in that document about account abstraction, the use of ecdsa_verify for those who bring their secp256k1 keys and edDsa verify on the other side (EVM). 

You would have an easier delivery by just doing Algorand sign-in instructions of trying to be EVM friendly. 

This design prevents in doing exactly what you set out to deliver, for example: 

> _SIWA implements a method to derive Ethereum-compatible addresses from Algorand public keys. This allows for interoperability with Ethereum-based systems and tools. The process involves decoding the Algorand address, hashing the public key with Keccak-256, and applying EIP-55 checksum rules to create an Ethereum-compatible address._

You simply apply a hashing function on a public key to try to get **another public key on a different curve!** You can't even know if the **result of the hash of a valid point?**

Even if you apply the secp256k1 curve equation to check if was so, you don't have a scalar for that "EVM"-compatible public key .  In ECC you need to have a scalar and mult on the generation point. How would we even sign with that "EVM public key"?

I would strongly discourage the current approach

headline-design commented:
> Any luck with Pera & Defly?

Working on it now

headline-design commented:
> > @SilentRhetoric Here is the link to the security document. This is my first time working with Latex so I'm happy to update as needed. https://github.com/headline-design/avmkit/blob/main/packages/siwa/security.tex
> 
> Thanks for the document, but i think this would need to be a technical document. It reads as an article.
> 
> Claiming interoperability on a SIWA scheme with EVM / SIWE is a bold endeavour and personally I don't think you demonstrated that. We should be reading in that document about account abstraction, the use of ecdsa_verify for those who bring their secp256k1 keys and edDsa verify on the other side (EVM).
> 
> You would have an easier delivery by just doing Algorand sign-in instructions of trying to be EVM friendly.
> 
> This design prevents in doing exactly what you set out to deliver, for example:
> 
> > _SIWA implements a method to derive Ethereum-compatible addresses from Algorand public keys. This allows for interoperability with Ethereum-based systems and tools. The process involves decoding the Algorand address, hashing the public key with Keccak-256, and applying EIP-55 checksum rules to create an Ethereum-compatible address._
> 
> You simply apply a hashing function on a public key to try to get **another public key on a different curve!** You can't even know if the **result of the hash of a valid point?**
> 
> Even if you apply the secp256k1 curve equation to check if was so, you don't have a scalar for that "EVM"-compatible public key . In ECC you need to have a scalar and mult on the generation point. How would we even sign with that "EVM public key"?
> 
> I would strongly discourage the current approach. It doesn't give AVM / EVM compatibility in any reasonable way

Thanks for the feedback on this. The Algorand address conforms to EIP-55 with this approach. EIP-55 is a standard check that most EVM applications preform. We pass the native Algorand address for signing validation and the EIP-55 compatible address for EIP-55 compatibility. Are you suggesting we make that part optional though? Since the feature is for EIP-55 compatibility? Right now EIP-55 compatibilty is a required part of the message scheme.

johnalanwoods commented:
> You simply apply a hashing function on a public key to try to get another public key on a different curve! You can't even know if the result of the hash of a valid point?

Extremely concerning.

johnalanwoods commented:
@headline-design it feels like you don't know what you're doing here.

do you have a proper architecture you can share?

I'm gravely concerned this will run in production.

headline-design commented:
@ehanoc here is the Pera and Defly template - https://siwa-connect.vercel.app. Repo template - https://github.com/headline-design/avmkit/tree/main/templates/siwa-connect. Also I've got the full documentation site up now at https://docs.siwa.org.

headline-design commented:
@johnalanwoods if you are referring to architectural diagram here is the basic diagram - https://docs.siwa.org/sign-in-with-algorand/quickstart-guide/sign-in-diagram. And here is the fullstack diagram - https://docs.siwa.org/general-information/siwa-overview/extended-siwa-diagram

headline-design commented:
> > You simply apply a hashing function on a public key to try to get another public key on a different curve! You can't even know if the result of the hash of a valid point?
> 
> Extremely concerning.

If the purpose of EIP-55 compatibility is confusing, I can remove it from the scheme. The derived address is not used for signing messages, only for EIP-55 compatibility.

headline-design commented:
Still have some work to do on docs. Added support now for Pera, Defly, Lute, and Kibisis wallets in SIWA Connect. Need to add new docs for that too and at least one more qc through all the templates to support the library updates.

ehanoc commented:
> Still have some work to do on docs. Added support now for Pera, Defly, Lute, and Kibisis wallets in SIWA Connect. Need to add new docs for that too and at least one more qc through all the templates to support the library updates.

I mean, have you spoken with Pera, Defly and others to adopt this? Writing code mentioning this "integrations" as it was defined in the proposal doesn't mean this will be available in the wallets.

ehanoc commented:
> Ok, I'm going to go ahead and remove the EIP-55 check. will take some time to update the docs but should have all the templates and the main library updated tonight.

This was part of the proposal and what the xGovs expected from the delivery point of view. The implementation doesn't seem correct to me though. If you want to move forward i would expect the AVM / EVM compatibility piece to be correctly delivered

headline-design commented:
> > Still have some work to do on docs. Added support now for Pera, Defly, Lute, and Kibisis wallets in SIWA Connect. Need to add new docs for that too and at least one more qc through all the templates to support the library updates.
> 
> I mean, have you spoken with Pera, Defly and others to adopt this? Writing code mentioning this "integrations" as it was defined in the proposal doesn't mean this will be available in the wallets.

I have updated the docs to include full documentation and support for all 4 wallets. I have also integrated all four wallets with the SIWA Connect template. I also updated the main SIWA package to provide validation based on provider. For Pera and Kibisis we verify with verifyBytes. For Defly and Lute we verify with a seperate verifyTransaction utility. All four wallets have been tested and validated. Ideally all wallets would support verifyBytes as this is the optimal verifcation message. I made a Defly Connect request for adding verifyBytes for that as well.

headline-design commented:
> > Ok, I'm going to go ahead and remove the EIP-55 check. will take some time to update the docs but should have all the templates and the main library updated tonight.
> 
> This was part of the proposal and what the xGovs expected from the delivery point of view. The implementation doesn't seem correct to me though. If you want to move forward i would expect the AVM / EVM compatibility piece to be correctly delivered

If you refer to the proposal and deliverables, the section on EVM-compatibility specifically refers to leveraging the Algorand blockchain for authentication. The Algostack EVM template fully demonstrates how seamlesly we can integrate Algorand authentication in an EVM application. The proposal makes no mention of signing EVM transaction or vice-versa, instead focusing specifically on standardizing how developers can add Algorand support to their apps. See here: "This is a critical feature for developers who build EVM-compatible applications and want to leverage the Algorand blockchain for authentication." Beyond this, I'm not entirely sure what you are looking for or expect to be delivered. If you refer to the abstract - "SIWA (Sign In With Algorand) is a decentralized authentication protocol that allows users to sign in to websites and applications using their Algorand wallet." Our methodolgy closely aligns with SIWE and EIP-4361. The entire EVM-integration feature here - "Feature 1: EVM Compatibility
SIWA will be modeled from the ground up to be compatible with SIWE, allowing developers to easily integrate SIWA into their existing Ethereum-based applications. This is a critical feature for developers who build Algorand applications with NextJS. This is a critical feature for developers who build EVM-compatible applications and want to leverage the Algorand blockchain for authentication." SIWA was modeled from the ground up to compatible with SIWE, and the Algostack EVM template demonstrates how developers can easilty integrate SIWA into Ethereum-based applications. 

If there are specific things you are looking for beyond this, I may need your help in implementing them as you may be looking for something specifc that I cannot entirely discern.

headline-design commented:
@ehanoc Also, here are the 4 individual wallet documentaion pages. Pera - https://docs.siwa.org/integrations/pera-wallet. Defly - https://docs.siwa.org/integrations/defly-wallet. Lute - https://docs.siwa.org/integrations/lute-wallet. Kibisis - https://docs.siwa.org/integrations/kibisis-wallet. Also I added an updated diagram to demonstrate how we handle verification for the different providers - https://docs.siwa.org/general-information/siwa-overview/sign-in-diagram.

ehanoc commented:
> > > Still have some work to do on docs. Added support now for Pera, Defly, Lute, and Kibisis wallets in SIWA Connect. Need to add new docs for that too and at least one more qc through all the templates to support the library updates.
> > 
> > 
> > I mean, have you spoken with Pera, Defly and others to adopt this? Writing code mentioning this "integrations" as it was defined in the proposal doesn't mean this will be available in the wallets.
> 
> I have updated the docs to include full documentation and support for all 4 wallets. I have also integrated all four wallets with the SIWA Connect template. I also updated the main SIWA package to provide validation based on provider. For Pera and Kibisis we verify with verifyBytes. For Defly and Lute we verify with a seperate verifyTransaction utility. All four wallets have been tested and validated. Ideally all wallets would support verifyBytes as this is the optimal verifcation message. I made a Defly Connect request for adding verifyBytes for that as well.

You don't generate valid ethereum addresses. They might look like having similar characters, but its not possible to find a valid public key on a different curve by hashing the previous one. For any EVM compatibility you would need to also be able to sign.

ehanoc commented:
> @ehanoc Also, here are the 4 individual wallet documentaion pages. Pera - https://docs.siwa.org/integrations/pera-wallet. Defly - https://docs.siwa.org/integrations/defly-wallet. Lute - https://docs.siwa.org/integrations/lute-wallet. Kibisis - https://docs.siwa.org/integrations/kibisis-wallet. Also I added an updated diagram to demonstrate how we handle verification for the different providers - https://docs.siwa.org/general-information/siwa-overview/sign-in-diagram.

I meant on the wallets themselves.

HabeebullahOpe reviewed (commented):
Progress

ehanoc commented:
> Progress

Besides being stalled this submission has serious implementation and technical design issues that weren't addressed.
